tcmur_cmd_handler.c

```diff
@@ -2308,6 +2308,361 @@ static int handle_inquiry(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+
+	switch (op) {
+	case TEST_UNIT_READY:
+		buf = test_unit_ready;
+		break;
+	case REQUEST_SENSE:
+		buf = request_sense;
+		break;
+	case FORMAT_UNIT:
+		buf = format_unit;
+		break;
+	case READ_6:
+	case WRITE_6:
+		buf = read_write_6;
+		break;
+	case INQUIRY:
+		buf = inquiry;
+		break;
+	case MODE_SELECT:
+		buf = mode_select;
+		break;
+	case RELEASE:
+	case RESERVE:
+		buf = reserve_release;
+		break;
+	case MODE_SENSE:
+		buf = mode_sense;
+		break;
+	case START_STOP:
+		buf = start_stop;
+		break;
+	case SEND_DIAGNOSTIC:
+		buf = send_diagnostics;
+		break;
+	case ALLOW_MEDIUM_REMOVAL:
+		buf = allow_medium_removal;
+		break;
+	case READ_CAPACITY:
+		buf = read_capacity;
+		break;
+	case READ_10:
+		buf = read_10;
+		break;
+	case WRITE_10:
+		buf = write_10;
+		break;
+	case WRITE_VERIFY:
+	case VERIFY_10:
+		buf = verify_10;
+		break;
+	case PRE_FETCH_10:
+		buf = pre_fetch_10;
+		break;
+	case SYNCHRONIZE_CACHE:
+		buf = synchronize_cache_10;
+		break;
+	case WRITE_SAME:
+		buf = write_same_10;
+		break;
+	case UNMAP:
+		buf = unmap;
+		break;
+	case MODE_SELECT_10:
+		buf = mode_select_10;
+		break;
+	case MODE_SENSE_10:
+		buf = mode_sense_10;
+		break;
+	case PERSISTENT_RESERVE_IN:
+		switch (sa) {
+		case PR_IN_READ_KEYS:
+		case PR_IN_READ_RESERVATION:
+		case PR_IN_REPORT_CAPABILITIES:
+		case PR_IN_READ_FULL_STATUS:
+			buf = persistent_reserve_in;
+			break;
+		}
+		break;
+	case PERSISTENT_RESERVE_OUT:
+		switch (sa) {
+		case PR_OUT_REGISTER:
+		case PR_OUT_RESERVE:
+		case PR_OUT_RELEASE:
+		case PR_OUT_CLEAR:
+		case PR_OUT_PREEMPT:
+		case PR_OUT_PREEMPT_AND_ABORT:
+		case PR_OUT_REGISTER_AND_IGNORE_EXISTING_KEY:
+		case PR_OUT_REGISTER_AND_MOVE:
+			buf = persistent_reserve_out;
+			break;
+		}
+		break;
+	case READ_16:
+		buf = read_16;
+		break;
+	case COMPARE_AND_WRITE:
+		buf = compare_and_write;
+		break;
+	case WRITE_16:
+	case ORWRITE_16:
+		buf = orwrite_16;
+		break;
+	case WRITE_VERIFY_16:
+	case VERIFY_16:
+		buf = verify_16;
+		break;
+	case PRE_FETCH_16:
+		buf = pre_fetch_16;
+		break;
+	case SYNCHRONIZE_CACHE_16:
+		buf = synchronize_cache_16;
+		break;
+	case WRITE_SAME_16:
+		buf = write_same_16;
+		break;
+	case SERVICE_ACTION_IN:
+		switch (sa) {
+		case SAI_READ_CAPACITY_16:
+			buf = read_capacity_16;
+			break;
+		case SAI_GET_LBA_STATUS:
+			buf = get_lba_status;
+			break;
+		}
+		break;
+	case REPORT_LUNS:
+		buf = report_luns;
+		break;
+	case MAINT_PROTOCOL_IN:
+		switch (sa) {
+		case MPI_REPORT_SUPPORTED_OPCODES:
+			buf = rep_sup_opcodes;
+			break;
+		}
+		break;
+	case READ_12:
+		buf = read_12;
+		break;
+	case VERIFY_12:
+	case WRITE_VERIFY_12:
+		buf = verify_12;
+		break;
+	case WRITE_12:
+		buf = write_12;
+		break;
+	}
+
+	if (buf) {
+		buf[0] = op;
+		return buf;
+	}
+
+	return usage;
+}
+
+int is_support_opcode( int opcode )
+{
+	switch( opcode ){
+	case INQUIRY:
+  		goto sup;
+	case TEST_UNIT_READY:
+		goto sup;
+	case WRITE_SAME:
+		goto sup;
+	case WRITE_SAME_16:
+		goto sup;
```

Comment:
Just do the fall through way:

```
case INQUIRY:
case TEST_UNIT_READY:
case WRITE_SAME:
....
    return true.
```

Also use bools.


What is this function for though? It is not clear to me why it only checks these specific commands?

Also, I think we will need a way for handler to report specific commands they support. For example, file_optical and file_zbc support some uncommon commands.